### PR TITLE
[Driver][SYCL][FPGA] Use opencl-aot for AOT when doing FPGA emulation

### PR DESCRIPTION
**Pull request by mdtoguchi**
When performing AOT compilation for FPGA, we call 'aoc' to perform
the offline compilation.  Update this behavior to use opencl-aot when
we know we are performing emulation instead of running on hardware.

When creating FPGA AOCX/AOCR archives when performing emulation, create
archives which contain .spv items as opposed to using the regular aocx
and aocr binaries.

The expected input type for AOCX/AOCR were considered for hardware
usage with FPGA.  In this part, we allow for the differentiation of
the type of archive created, whether it be emulation or hardware.
Depending on how these archives are created, we make sure we take
the proper flow to process the contents.

Also perform some cleanup in the needed triple arch representation
for fpga, where we didn't need to define all of the different types
we were wrapping for FPGA specific binaries.

### CODE REVIEW

**Comment by keryell** [comment]
What is the difference between `aocx` and `aocr`?
I am trying to understand the difference with the end of section https://github.com/triSYCL/sycl/blob/sycl/unified/master/sycl/doc/GettingStartedXilinxFPGA.md#small-examples

**Comment by mdtoguchi** [comment]
> What is the difference between `aocx` and `aocr`?
> I am trying to understand the difference with the end of section https://github.com/triSYCL/sycl/blob/sycl/unified/master/sycl/doc/GettingStartedXilinxFPGA.md#small-examples

`aocx` is considered a final image, while `aocr` is an early image which can be processed again.  @meiyacha, can you provide details of the differences?

**Comment by meiyacha** [comment]
> > What is the difference between `aocx` and `aocr`?
> > I am trying to understand the difference with the end of section https://github.com/triSYCL/sycl/blob/sycl/unified/master/sycl/doc/GettingStartedXilinxFPGA.md#small-examples
> 
> `aocx` is considered a final image, while `aocr` is an early image which can be processed again. @meiyacha, can you provide details of the differences?

That's correct. `aocx` is the image after the hw compile and produces the FPGA bitstream; `aocr` is partial hw compile that gives us enough info to provide an estimate report to the user to verify their design with a shorter turnaround time for iteration, but needs to compiled to `aocx` to produce the FPGA bitstream. The `aocr` stage is at the order of minutes while `aocx` is at the order of hours.

File: clang/include/clang/Driver/Driver.h [hunk 1]
```diff
@@ -634,6 +634,11 @@ class Driver {
 
   void setOffloadStaticLibSeen() { OffloadStaticLibSeen = true; }
 
+  /// FPGA Emulation Mode.  By default, this is true due to the fact than
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/include/clang/Driver/Driver.h]: Updated.

**Review by bader — APPROVED** [review]
llvm/* changes look good to me.

**Comment by mdtoguchi** [comment]
@AGindinson , any update on review?  I'll update to fix the conflicts.

**Review by AGindinson — COMMENTED** [review]
The design LGTM, I'll revise the code more closely later today.

**Review by premanandrao — APPROVED** [review]
The changes in the couple of files pertaining to FE, look good to me.

File: clang/include/clang/Driver/ToolChain.h [hunk 2]
```diff
@@ -339,6 +339,12 @@ class ToolChain {
                      Action::OffloadKind DeviceOffloadKind,
                      SmallVectorImpl<llvm::opt::Arg *> *AllocatedArgs) const;
 
+  /// TranslateBackendTargetArgs - populate an argument list with the specific
+  /// options that are determined to be passed along to an offline backend
+  /// target compiler.
+  void TranslateBackendTargetArgs(const llvm::opt::ArgList &Args,
+                                  llvm::opt::ArgStringList &CmdArgs) const;
```
Comment (AGindinson) [inline clang/include/clang/Driver/ToolChain.h:346]: Could you please clarify why porting this function to the base toolchain class is required? I'm sure there's rationale that I simply haven't gathered from the code by this stage.

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -4709,6 +4728,46 @@ class OffloadingActionBuilder final {
     return C.MakeAction<OffloadAction>(HDep, DDeps);
   }
 
+  // Update Input action to reflect FPGA device archive specifics based
+  // on archive contents.
+  bool updateInputForFPGA(Action *&A, const Arg *InputArg,
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:4733]: Wouldn't a simple pointer type do the trick as well?
```suggestion
  bool updateInputForFPGA(Action *A, const Arg *InputArg,
```

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 4]
```diff
@@ -308,6 +358,19 @@ void SYCL::fpga::BackendCompiler::ConstructJob(
...
 
+  // Grab the -Xsycl-target* options.
+  const toolchains::SYCLToolChain &TC =
+      static_cast<const toolchains::SYCLToolChain &>(getToolChain());
+  ArgStringList TargetArgs;
+  TC.TranslateBackendTargetArgs(Args, TargetArgs);
+
+  // When performing emulation compilations for FPGA AOT, we want to use
+  // opencl-aot instead of aoc.
+  if (C.getDriver().isFPGAEmulationMode()) {
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp:369]: How would we have reached this block if the condition was not true?

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:369]: We call the BE compiler for FPGA for `spir64_fpga` regardless of emulation mode or not.

File: clang/lib/Driver/Driver.cpp (hunk 3 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:4738]: I don't think so, I'm passing by reference as I'm creating a new action and replacing if needed.

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 4 above)
Comment (AGindinson) [reply clang/lib/Driver/ToolChains/SYCL.cpp:369]: Ah, of course. I realized this on the design review stage, but then it somehow slipped my mind when the Toolchain time came :D

File: clang/include/clang/Driver/Driver.h [hunk 5]
```diff
@@ -641,6 +641,11 @@ class Driver {
 
   void setOffloadStaticLibSeen() { OffloadStaticLibSeen = true; }
 
+  /// FPGA Emulation Mode.  By default, this is true due to the fact that
+  /// an external option setting is required to target hardware.
+  bool FPGAEmulationMode = true;
+  void setFPGAEmulationMode(bool Mode) { FPGAEmulationMode = Mode; }
```
Comment (AGindinson) [inline clang/include/clang/Driver/Driver.h:647]:
```suggestion
  void setFPGAEmulationMode(bool IsEmulation) { FPGAEmulationMode = IsEmulation; }
```

File: clang/lib/Driver/Driver.cpp [hunk 6]
```diff
@@ -4709,6 +4733,46 @@ class OffloadingActionBuilder final {
...
+    }
+
+    SmallVector<std::pair<types::ID, bool>, 4> FPGAAOCTypes = {
+        {types::TY_FPGA_AOCX, false},
+        {types::TY_FPGA_AOCR, false},
+        {types::TY_FPGA_AOCX_EMU, true},
+        {types::TY_FPGA_AOCR_EMU, true}};
+    for (const auto &ArchiveType : FPGAAOCTypes) {
+      bool binaryFound = hasFPGABinary(C, InputName, ArchiveType.first);
+      if (binaryFound && ArchiveType.second == IsFPGAEmulation) {
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:4766]: Suggest renaming as per [the guidelines](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly):
```suggestion
      bool BinaryFound = hasFPGABinary(C, InputName, ArchiveType.first);
      if (BinaryFound && ArchiveType.second == IsFPGAEmulation) {
```

File: clang/lib/Driver/Driver.cpp [hunk 7]
```diff
@@ -4709,6 +4733,46 @@ class OffloadingActionBuilder final {
...
+  // on archive contents.
+  bool updateInputForFPGA(Action *&A, const Arg *InputArg,
+                          DerivedArgList &Args) {
+    std::string InputName = InputArg->getAsString(Args);
+    bool IsFPGAEmulation = C.getDriver().isFPGAEmulationMode();
+    // Only check for FPGA device information when using fpga SubArch.
+    if (A->getType() == types::TY_Object && isObjectFile(InputName))
+      return true;
+
+    auto ArchiveTypeMismatch = [&](bool EmitDiag) {
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:4746]: Wonder if a micro-optimization this is possible:
```suggestion
    auto ArchiveTypeMismatch = [&C, &InputName](bool EmitDiag) {
```

File: clang/lib/Driver/Driver.cpp (hunk 3 above)
Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp:4738]: Took me a while to grasp, but clear enough now, thanks! It's essentially the same mechanics as with the `addHostDependenceToDeviceActions(...)` caller.

**Review by AGindinson — COMMENTED** [review]
The latest ToolChain-related update has really simplified the changes. The only comments left on my part are minor stylistic nits!

**Review by AGindinson — DISMISSED** [review]
Great job, thanks!

File: clang/lib/Driver/Driver.cpp [hunk 8]
```diff
@@ -1377,6 +1377,22 @@ Compilation *Driver::BuildCompilation(ArrayRef<const char *> ArgList) {
   // Populate the tool chains for the offloading devices, if any.
   CreateOffloadingDeviceToolChains(*C, Inputs);
 
+  // Determine FPGA emulation status.
+  if (C->hasOffloadToolChain<Action::OFK_SYCL>()) {
+    auto SYCLTCRange = C->getOffloadToolChains<Action::OFK_SYCL>();
+    ArgStringList TargetArgs;
+    const ToolChain *TC = SYCLTCRange.first->second;
+    const toolchains::SYCLToolChain *SYCLTC =
+                static_cast<const toolchains::SYCLToolChain *>(TC);
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:1386]: This bit seems to be annoying clang-format:
```suggestion
    const toolchains::SYCLToolChain *SYCLTC =
           static_cast<const toolchains::SYCLToolChain *>(TC);
```

**Review by AGindinson — APPROVED** [review]

**Review by premanandrao — APPROVED** [review]
FE part looks good to me.

**Comment by mdtoguchi** [comment]
@bader, one final (hopefully) look?

**Review by bader — APPROVED** [review]

**Pull request merged**